### PR TITLE
Ports: Add elfutils (and a bunch of other dependencies...)

### PR DESCRIPTION
**Pull request by supercomputer7**
Relies on #19728.

### CODE REVIEW

**Comment by supercomputer7** [comment]
@timschumi Now here's the real deal - I've got this to work, but it is really clunky (some utilities need more testing) - we can't compile all of them so 5 of them are disabled for now. It seems like GNU binutils give us pretty much all we need, except for the DWARF code, so I'll give you the decision on whether to close this or continue working on this.

Either way, it was the hardest port I've ever pulled and managed to get into a working state, but it was quite fun too :)

File: Ports/nnn/patches/0003-Disable-broken-features.patch [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
++O_NOMOUSE := 1  # no mouse support
+ O_NOBATCH := 0  # no built-in batch renamer
+ O_NOFIFO := 0  # no FIFO previewer support
+ O_CTX8 := 0  # enable 8 contexts
+@@ -26,7 +26,7 @@ O_QSORT := 0  # use Alexey Tourbin's QSORT implementation
+ O_BENCH := 0  # benchmark mode (stops at first user input)
+ O_NOSSN := 0  # disable session support
+ O_NOUG := 0  # disable user, group name in status bar
+-O_NOX11 := 0  # disable X11 integration
++O_NOX11 := 1  # disable X11 integration
```
Comment (ADKaster) [inline Ports/nnn/patches/0003-Disable-broken-features.patch:29]: surely we can set these with configopts or by passing an argument to `make`? `make NO_X11=1` or some such?

File: Userland/Libraries/LibC/crt0.cpp [hunk 2]
```diff
@@ -45,6 +45,8 @@ int _entry(int argc, char** argv, char** env)
 
     s_global_initializers_ran = true;
 
+    program_invocation_short_name = argv[0];
```
Comment (ADKaster) [inline Userland/Libraries/LibC/crt0.cpp:48]: This is not the same as argv[0]. 

https://man7.org/linux/man-pages/man3/program_invocation_name.3.html

It also belongs in `errno.h`, not in `stdio.h`

>  program_invocation_short_name contains the basename component of name that was used to invoke the calling program.  That is, it is the same value as program_invocation_name, with all text up to and including the final slash (/), if any, removed.

File: Ports/argp-standalone/package.sh [hunk 3]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port='argp-standalone'
+version='1.0.2'
+useconfigure='true'
+files="https://github.com/tom42/argp-standalone/archive/refs/tags/v1.0.2.zip argp.zip 75b624a5e9e89d5e7527400a36c9b434e3f7eeaf3e8c19baf018ea4be3e250f9"
+auth_type='sha256'
+
+configure() {
+    run cmake "${configopts[@]}" .
```
[the remark this answers is not in the document]
Comment (timschumi) [reply Ports/argp-standalone/package.sh:9]: No, configopts is empty by default. CMake-based ports need to add `CMAKE_TOOLCHAIN_FILE` manually.

\*Adds one to the tally mark for "remove configopts entirely"\*

File: Userland/Libraries/LibC/stdio.cpp [hunk 4]
```diff
@@ -728,19 +728,29 @@ int ungetc(int c, FILE* stream)
...
     u8 byte = ch;
-    ScopedFileLock lock(stream);
     size_t nwritten = stream->write(&byte, 1);
     if (nwritten == 0)
         return EOF;
     VERIFY(nwritten == 1);
     return byte;
 }
 
+int putc_unlocked(int ch, FILE* stream)
```
Comment (timschumi) [inline Userland/Libraries/LibC/stdio.cpp:773]: This actually does have a POSIX manpage.

File: Userland/Libraries/LibC/err.h [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+     __attribute__ ((__format__ (__printf__, 1, 0)));
+
+void err (int __status, const char *__format, ...)
+     __attribute__ ((noreturn, __format__ (__printf__, 2, 3)));
+void verr (int __status, const char *__format, va_list)
+     __attribute__ ((noreturn, __format__ (__printf__, 2, 0)));
+void errx (int __status, const char *__format, ...)
+     __attribute__ ((noreturn, __format__ (__printf__, 2, 3)));
+void verrx (int __status, const char *, va_list)
+     __attribute__ ((noreturn, __format__ (__printf__, 2, 0)));
```
Comment (timschumi) [inline Userland/Libraries/LibC/err.h:31]: Where is this copied from? Both the argument naming as well as the general formatting don't match what we usually use. :yakthonk:

File: Ports/argp-standalone/package.sh [hunk 6]
```diff
@@ -0,0 +1,16 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port='argp-standalone'
+version='1.0.2'
+useconfigure='true'
+files="https://github.com/tom42/argp-standalone/archive/refs/tags/v1.0.2.zip argp.zip 75b624a5e9e89d5e7527400a36c9b434e3f7eeaf3e8c19baf018ea4be3e250f9"
```
Comment (timschumi) [inline Ports/argp-standalone/package.sh:5]: This could use `${version}`.

File: Ports/elfutils/package.sh [hunk 7]
```diff
@@ -0,0 +1,12 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+port='elfutils'
+version='0.186'
+useconfigure='true'
+use_fresh_config_sub='true'
+config_sub_paths=('config/config.sub')
+files="http://archive.ubuntu.com/ubuntu/pool/main/e/elfutils/elfutils_${version}.orig.tar.bz2 elfutils_${version}.orig.tar.bz2 7f6fb9149b1673d38d9178a0d3e0fb8a1ec4f53a9f4c2ff89469609879641177"
+auth_type='sha256'
+depends=('zlib' 'libfts' 'argp-standalone' 'musl-obstack' 'curl' 'libarchive' 'sqlite' 'gettext')
```
Comment (timschumi) [inline Ports/elfutils/package.sh:9]: As usual, one line for each (and sorted).

File: Userland/Libraries/LibC/ar.h [hunk 8]
```diff
@@ -0,0 +1,47 @@
+/* Header describing `ar' archive file format.
+   Copyright (C) 1996-2023 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
```
Comment (timschumi) [inline Userland/Libraries/LibC/ar.h:3]: Just to note that down somewhere, writing our own version of ar.h should be trivial if the contents are not protectable (and if they are, doing something like this is out of the question anyways).

File: Userland/Libraries/LibC/stdio_locking_ext.h [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+
+enum FSETLOCKING {
+  FSETLOCKING_QUERY = 0,
+  FSETLOCKING_INTERNAL,
+  FSETLOCKING_BYCALLER
+};
+
+#define FSETLOCKING_QUERY	FSETLOCKING_QUERY
+#define FSETLOCKING_INTERNAL	FSETLOCKING_INTERNAL
+#define FSETLOCKING_BYCALLER	FSETLOCKING_BYCALLER
```
Comment (timschumi) [inline Userland/Libraries/LibC/stdio_locking_ext.h:22]: Judging by the manpage on die.net, all of this should be in the normal `stdio_ext.h` file.

File: Ports/elfutils/package.sh [hunk 10]
```diff
@@ -8,5 +8,36 @@ files="http://archive.ubuntu.com/ubuntu/pool/main/e/elfutils/elfutils_${version}
...
 )
+
+pre_configure() {
+    # Rebuild after patching configure.ac to support serenity host.
+
+    # `automake` may exit with a warning about how there is a mismatch
+    # between the versions of autoconf and automake that were previously
+    # used to generate aclocal and specifed in configure.ac.
+    run automake --add-missing || true
+    run autoreconf
```
Comment (timschumi) [inline Ports/elfutils/package.sh:24]: The tarball arrives preconfigured, so we should just patch the existing `configure` file. Everything else risks many more compatibility issues.

**Review by timschumi — CHANGES_REQUESTED** [review]

**Pull request closed without merging**